### PR TITLE
[networking] Remove dot segments during URL normalization

### PR DESCRIPTION
**Pull request by coletdjnz**
This implements [RFC3986 5.2.4](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2.4) `remove_dot_segments` during the URL normalization process, particularly for the urllib handler.

Closes https://github.com/yt-dlp/yt-dlp/issues/3355, https://github.com/yt-dlp/yt-dlp/issues/6526

This is adapted from the remove_dot_segments pseudo-code in the RFC and some inspiration from urllib3/rfc396 libraries (though it came out very close to them).

I have also renamed `escape_url` to `normalize_url` to better represent what it is doing, and moved these functions to `utils.networking`.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [X] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🚚🔧🧪

This pull request refactors and improves the URL handling functions in yt-dlp. It introduces a new function `normalize_url` that removes dot segments and escapes non-ASCII characters in a URL, and uses it in various modules. It also adds a new test case for the dot segment removal algorithm.

> _To handle URLs with more care_
> _They moved and renamed `escape_url`_
> _Now `normalize_url` is the tool_
> _That removes dot segments and escapes_
> _And makes the networking module cool_

### Walkthrough
*  Refactor and move the functions for escaping and normalizing URLs to `yt_dlp/utils/networking.py` and implement the RFC 3986 5.2.4 algorithm for removing dot segments from a path ([link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-b6d91e8afdfc47fb2b59e37d22d3621b79d252bc5a8205edc2b80530dcceecc2R118-R155), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-feda8f56946dc048e754111850baaef0eec4b9f8bbc2d3f04b1a785626ea5c0eL2467-L2483), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-0a7c9732b38802b4a7b67fab5d9ae6a6ff1722586503b945230e59862ac24db0R11-R12), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-0a7c9732b38802b4a7b67fab5d9ae6a6ff1722586503b945230e59862ac24db0L200-R202), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-ead7fca4af365bb8c89c1a0257a027147d0533eb2e0dfb395d08a9a12e58d4abL29-R31), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-ead7fca4af365bb8c89c1a0257a027147d0533eb2e0dfb395d08a9a12e58d4abL371-R370), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-c7b2bfc3ad326c23db07694e9137f86e58e526681dab17e7680a968a7190e232L44-R45), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-c7b2bfc3ad326c23db07694e9137f86e58e526681dab17e7680a968a7190e232L182-R183), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-c7b2bfc3ad326c23db07694e9137f86e58e526681dab17e7680a968a7190e232L215-R216), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7L36), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7R44), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7L1311-R1311), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-c042473d560a04c6fd504c3cb3ae2c9a56a27e42458132a265a3ffda0d60a5c7L1320-R1320), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-1780a2492dac55d0b4c5ff13cd3cf852a72cdac8995e17b10c87eeb678b3179eL50-L51), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-1780a2492dac55d0b4c5ff13cd3cf852a72cdac8995e17b10c87eeb678b3179eL135-R138), [link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-1780a2492dac55d0b4c5ff13cd3cf852a72cdac8995e17b10c87eeb678b3179eL936-R979))
* Add a new handler for the path `/redirect_dotsegments` in the `HTTPTestRequestHandler` class, which sends a 301 redirect response with a location header that contains dot segments ([link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-e92498eb46ad8369b16aed0d65fa7c8d56c81b2d595f6510981c6cfb0c0f7e78R176-R181))
* Add a new test case for the `remove_dot_segments` and `normalize_url` functions, which uses the `TestHTTPRequestHandler` class and the `validate_and_send` function to send two requests: one to the path `/a/b/./../../headers` and one to the path `/redirect_dotsegments`. The test case asserts that both requests result in a 200 status and a final URL of `/headers` ([link](https://github.com/yt-dlp/yt-dlp/pull/7662/files?diff=unified&w=0#diff-e92498eb46ad8369b16aed0d65fa7c8d56c81b2d595f6510981c6cfb0c0f7e78R365-R379))

</details>

### CODE REVIEW

**Comment by dirkf** [comment]
Does this handle the problem URL`https://reoa92d.com/../uploaded/1649416469.mp4#t=0.1` from the original #3355 report? I thought this was against RFCs but maybe there's new logic (ISTR checking the WHAT spec for this when I made up the patch in that issue).

**Comment by coletdjnz** [comment]
> Does this handle the problem URL`https://reoa92d.com/../uploaded/1649416469.mp4#t=0.1` from the original #3355 report? I thought this was against RFCs but maybe there's new logic (ISTR checking the WHAT spec for this when I made up the patch in that issue).

Yes it works (though I think the domain is dead), and pretty sure it is not against standard:

RFC 3986 5.2.4 Step A:
> A.  If the input buffer begins with a prefix of "../" or "./",
           then remove that prefix from the input buffer; otherwise,

Step C is similar for `/../`

File: test/test_utils.py [hunk 1]
```diff
@@ -933,24 +936,46 @@ def test_escape_rfc3986(self):
...
+            normalize_url('http://тест.рф/абв?абв=абв#абв'),
             'http://xn--e1aybc.xn--p1ai/%D0%B0%D0%B1%D0%B2?%D0%B0%D0%B1%D0%B2=%D0%B0%D0%B1%D0%B2#%D0%B0%D0%B1%D0%B2'
         )
-        self.assertEqual(escape_url('http://vimeo.com/56015672#at=0'), 'http://vimeo.com/56015672#at=0')
+        self.assertEqual(normalize_url('http://vimeo.com/56015672#at=0'), 'http://vimeo.com/56015672#at=0')
+
+        self.assertEqual(normalize_url('http://www.example.com/../a/b/../c/./d.html'), 'http://www.example.com/a/c/d.html')
+
+    def test_remove_dot_segments(self):
+        # XXX: we really should use pytest with parametrize here for my sanity
```
Comment (pukkandan) [inline test/test_utils.py:961]:
```suggestion
```
This'd have to be done for the whole file

File: yt_dlp/utils/networking.py [hunk 2]
```diff
@@ -115,3 +115,41 @@ def clean_headers(headers: HTTPHeaderDict):
...
+    output = []
+    segments = path.split('/')
+    for s in segments:
+        if s == '.':
+            continue
+        elif s == '..':
+            if output:
+                output.pop()
+        else:
+            output.append(s)
```
Comment (pukkandan) [inline yt_dlp/utils/networking.py:133]: This can be significantly simplified with regex, but I'll let it be if you prefer this way

**Review by pukkandan — APPROVED** [review]

**Pull request merged**